pkg/controllers/corecluster/corecluster.go

```diff
@@ -0,0 +1,145 @@
+func (r *CoreClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := ctrl.NewControllerManagedBy(mgr).
+		Named(controllerName).
+		For(r.Cluster).
+		Complete(r); err != nil {
+		return fmt.Errorf("failed to create controller: %w", err)
+	}
+
+	return nil
+}
+
+// Reconcile reconciles the core cluster object for the openshift-cluster-api namespace.
+func (r *CoreClusterReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", req.Namespace, "name", req.Name)
+
+	// Still have to patch the condition onto a newly created cluster
+	cluster, err := CreateCoreCluster(r.Client, r.Cluster.Name, r.Infra.Kind)
+	if err != nil {
+		logger.Info("failed to create new core cluster.")
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, cluster); err != nil && !errors.IsNotFound(err) {
+		return ctrl.Result{}, fmt.Errorf("failed to get core cluster: %w", err)
+	}
+
+	if !cluster.DeletionTimestamp.IsZero() {
+		if err := r.SetStatusAvailable(ctx, ""); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	logger.V(1).Info("Reconciling core cluster")
+	defer logger.V(1).Info("Finished reconciling core cluster")
+
+	clusterCopy := cluster.DeepCopy()
+
+	// Check if the Control Plane Initialized condition is set on an already existing cluster.
+	// If it's not, set it.
+	if conditions.Get(cluster, clusterv1.ControlPlaneInitializedCondition) == nil {
+		conditions.MarkTrue(cluster, clusterv1.ControlPlaneInitializedCondition)
+
+		patch := client.MergeFrom(clusterCopy)
+
+		isRequired, err := util.IsPatchRequired(cluster, patch)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to check if patch required: %w", err)
+		}
+
+		if isRequired {
+			if err := r.Status().Patch(ctx, cluster, patch); err != nil {
+				return ctrl.Result{}, fmt.Errorf("unable to update core cluster status: %w", err)
+			}
+		}
+
+		if err := r.SetStatusAvailable(ctx, ""); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to set status available: %w", err)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// CreateCluster creates a cluster with the given name and returns the cluster object.
+func CreateCoreCluster(cl client.Client, clusterName, infraClusterKind string) (*clusterv1.Cluster, error) {
+	CAPINamespace := "openshift-cluster-api"
+
+	cluster := &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      clusterName,
+			Namespace: CAPINamespace,
+		},
+		Spec: clusterv1.ClusterSpec{
+			InfrastructureRef: &corev1.ObjectReference{
+				APIVersion: "infrastructure.cluster.x-k8s.io/v1beta1",
+				Kind:       infraClusterKind,
+				Name:       clusterName,
+				Namespace:  CAPINamespace,
+			},
+		},
+	}
+
+	if infraClusterKind == "VSphereCluster" {
+		host, port, err := GetControlPlaneHostAndPort(cl)
```

Comment:
I'm a bit confused on the checking of the host and port for VSphereClusters - [is it necessary to take this function from the e2e tests](https://github.com/openshift/cluster-capi-operator/blob/ec9120bb11ace966a9046ad8e698216f8abf5c42/e2e/framework/util.go#L12), or is there a more succinct version I can do here?

Reply:
Let's not branch on infraClusterKind from here.
Let's do it in the same way as https://github.com/openshift/cluster-capi-operator/blob/ec9120bb11ace966a9046ad8e698216f8abf5c42/pkg/controllers/infracluster/aws.go#L49-L61

Reply:
Thanks, I've updated it to look more like that now